<a href="https://colab.research.google.com/github/Sourabhpatel1/Capstone-Project-EDA-AlmaBatter/blob/main/Copy_of_Telecom_Churn_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Import required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from collections import Counter
%matplotlib inline

In [2]:
# Mounting Google Drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
# loading data
churn_data = pd.read_csv("/content/drive/MyDrive/Copy of Telecom Churn.csv")

In [4]:
# checking null values and column info
churn_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

<p>There are no null Values in the Columns.</p> 
<p>Thus there is no need to handle null values.</p>
<p> There are 16 numerical columns (8 float 64 and 8 int64), 3 object (text/categorical) columns and 1 column with boolean (True/False) datatype </p> 

 

In [5]:
# Checking the number of columns

print(churn_data.columns)
print(f"\n There are {len(churn_data.columns)} columns in the dataset")

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')

 There are 20 columns in the dataset


In [6]:
# checking the head
churn_data.head()

State  Account length  ...  Customer service calls  Churn
0    KS             128  ...                       1  False
1    OH             107  ...                       1  False
2    NJ             137  ...                       0  False
3    OH              84  ...                       2  False
4    OK              75  ...                       3  False

[5 rows x 20 columns]

In [7]:
# checking the tail
churn_data.tail()

State  Account length  ...  Customer service calls  Churn
3328    AZ             192  ...                       2  False
3329    WV              68  ...                       3  False
3330    RI              28  ...                       2  False
3331    CT             184  ...                       2  False
3332    TN              74  ...                       0  False

[5 rows x 20 columns]

In [8]:
# Getting Overview of the Dataset
churn_data.describe()

Account length    Area code  ...  Total intl charge  Customer service calls
count     3333.000000  3333.000000  ...        3333.000000             3333.000000
mean       101.064806   437.182418  ...           2.764581                1.562856
std         39.822106    42.371290  ...           0.753773                1.315491
min          1.000000   408.000000  ...           0.000000                0.000000
25%         74.000000   408.000000  ...           2.300000                1.000000
50%        101.000000   415.000000  ...           2.780000                1.000000
75%        127.000000   510.000000  ...           3.270000                2.000000
max        243.000000   510.000000  ...           5.400000                9.000000

[8 rows x 16 columns]

<h1><strong> Dataset in a Glance</strong><h1>

> The describe function without any parameters
returns the description of only numerical columns. For analysing the Categorical and Boolean data the describe function has to be called with specific parameters, which has been done in the next cell

---


<ol>
  <li>Average account lenghth is approximately 101 days.</li>
  <li>Although the average number of calls in all three timeblocks i.e. Day, Evening, Night is 100 but the average duration of calls varies significantly with time of the day.</li>
  <li>Average call duration during day is 179.77 minutes, during evening it is 200 minutes on an average and 200 minutes on an average during the night.</li>
  <li>The mean value of number of voicemails per customer is 8 on an average, however, the standard deviation is very high suggesting that the number of customers availing voicemail serveces are very few and sporadic.</li>
  <li> Average international minutes is around 10 minutes per customer and average international call per customer is 4 suggesting that very few customers call internationally.</li>
</ol>


In [9]:
# Getting overview of the categorical and boolean data columns

churn_data.describe(include=["object", "bool"])

State International plan Voice mail plan  Churn
count   3333               3333            3333   3333
unique    51                  2               2      2
top       WV                 No              No  False
freq     106               3010            2411   2850

<h1><strong>Categorical Data at a glance</strong></h1>
<ol>
  <li>There are 51 unique states in the datatset with State Code WV having the most number of subscribers i.e. 106 subscribers.</li>
  <li>Out of the 3333 subscribers 3010 subscribers do not have a International Plan.</li>
  <li>Out of 3333 subscribers 2411 dont have a voice mail plan.</li>
  <li> Out of 3333 subscribers 483 subscribers have stoppped using the services of the company.</li>
  <li> 14.50 percent of subscribers have stooped using the services of the comapny thus churn rate of the comapny is 14.50 % which is very high.</li>
</ol>


In [10]:
# Making extra columns for finding out the per minute tariff for Day, Evening and Night charges.

churn_data['Day tariff'] = round(churn_data['Total day charge']/churn_data['Total day minutes'],2)
churn_data['Eve tariff'] = round(churn_data['Total eve charge']/churn_data['Total eve minutes'],2)
churn_data['Night tariff'] = round(churn_data['Total night charge']/churn_data['Total night minutes'],2)
churn_data

State  Account length  Area code  ... Day tariff Eve tariff  Night tariff
0       KS             128        415  ...       0.17       0.09          0.04
1       OH             107        415  ...       0.17       0.09          0.05
2       NJ             137        415  ...       0.17       0.08          0.05
3       OH              84        408  ...       0.17       0.08          0.04
4       OK              75        415  ...       0.17       0.09          0.04
...    ...             ...        ...  ...        ...        ...           ...
3328    AZ             192        415  ...       0.17       0.09          0.05
3329    WV              68        415  ...       0.17       0.09          0.05
3330    RI              28        510  ...       0.17       0.09          0.05
3331    CT             184        510  ...       0.17       0.09          0.04
3332    TN              74        415  ...       0.17       0.08          0.04

[3333 rows x 23 columns]

In [41]:
# Checking the average number of customer service calls between the people who have opted out of service.
churn_service_calls = pd.DataFrame(churn_data.groupby("Churn")["Customer service calls"].mean())
churn_service_calls

Customer service calls
Churn                        
False                1.449825
True                 2.229814

In [51]:
# Ratio of Customer service calls between subscribers who continue to use the service and who have stopped using the operators services.
churn_false = float(churn_service_calls.loc[False])
churn_true = float(churn_service_calls.loc[True])

churn_service_call_ratio = churn_true/churn_false
round(churn_service_call_ratio, 2)

1.54

In [58]:
service_calls_churn = pd.DataFrame(churn_data.loc[churn_data["Churn"] == True, "Customer service calls"])
service_calls_churn.describe()

Customer service calls
count              483.000000
mean                 2.229814
std                  1.853275
min                  0.000000
25%                  1.000000
50%                  2.000000
75%                  4.000000
max                  9.000000

In [60]:
service_calls_nonchurn = pd.DataFrame(churn_data.loc[churn_data["Churn"] == False, "Customer service calls"])
service_calls_nonchurn.describe()

Customer service calls
count             2850.000000
mean                 1.449825
std                  1.163883
min                  0.000000
25%                  1.000000
50%                  1.000000
75%                  2.000000
max                  8.000000

<h1><strong>Number of Service Calls<strong></h1>
<ol>
  <li>The subscribers who have stopped using the service of the operator have called 1.54 times more than the subscribers who continue to use the services of the operator.</li>
  <li>The high numbers of service calls may be due to poor quality of service or due to lack of attentive and friendly call center executives.</li>
</ol>